rfc-149-switch-to-per-page-asset-loading.md

```diff
@@ -0,0 +1,209 @@
+
+### HTTP requests
+
+The number of requests per page would go up - this shouldn't be a problem as [GOV.UK has HTTP/2 enabled][http2], and is due to have [HTTP/3][http3] enabled in the near future. During the past month 99.5% of tracked visitors used browsers that support HTTP/2, and 66.2% of visitors used browsers that supported HTTP/3.
+
+Of the top 100 most visited pages on GOV.UK, the average number of components on a page is 19. The smallest number of components on a page is 9, and the largest number of components is 25. It's worth noting that most - but not all - components need a stylesheet (72 out of 78 components have a stylesheet) and not every component needs JavaScript (24 out of 78 components have a JavaScript file) - so adding a component to a page doesn't always mean two more requests are made.
+
+It's worth noting that this change will lead to slower load times for browsers that only support HTTP/1.1 - even when considering the reduction in page size. ([More information on HTTP/1.1's limitations][http_limitations]). This will impact Internet Explorer 11 on Windows 8 and below; the first versions of Chrome, Firefox, and Opera that supported HTTP/2 were released in early 2015.
+
+This balance between making the site load faster for a majority of users is not just a numbers game - the users who are most likely to be using older browsers are those that are likely to least afford the upgrade cost for a newer device, and they're likely to be the users who most need the information on GOV.UK.
+
+Whilst this proposed change will make things slower for users whose browser doesn't support HTTP/2 or HTTP/3, it won't prevent pages from loading. This proposed change should be accepted because it will dramatically improve page performance for almost all visitors without blocking access for a small number of visitors.
+
+### Non-rendering applications that use the components gem
+
+There are applications that use the components gem but don't use Static and Slimmer - for example, publishing applications and the Govspeak gem.
+
+These changes should be optional - even if this is the default behaviour, it should be able to be turned off at a component by component level and at an application level. This could be done with a helper method that checks both if the component has already been used, and if adding the component assets are allowed:
+
+```ruby
+# component_asset_helper.rb - pseudo code example:
+
+def component_assets_allowed (component_name)
+  # check:
+  #  - if the component has already had the assets added to the page
+  # - if the component assets loading method is allowed at an application level
+  # - if the individual component has used an option to turn individual assets on or off
+  @components_in_use.include?(component_name) || ENV.component_asset_loading !== "bulk" || use_individual_assets === true
```

Comment:
Implementation detail, but I do wonder where this `use_individual_assets` would get set, and where this helper tool should live.
We'll want to spike how this might work - but fair enough if that's out of scope for the RFC itself.

Reply:
I agree that this should be a spike of work - but I think that this is outside the scope of this RFC, as it's getting closer to _how_ this should be implemented rather than _why_ this should be implemented.